jgrapht-core/src/main/java/org/jgrapht/intervalgraph/CenteredIntervalTree.java

```diff
@@ -0,0 +1,110 @@
+package org.jgrapht.intervalgraph;
+
+import org.jgrapht.Graph;
+import org.jgrapht.intervalgraph.interval.Interval;
+import org.omg.SendingContext.RunTime;
+
+import java.util.*;
+
+public class CenteredIntervalTree<T extends Comparable<T>> implements IntervalGraph<T>{
+    private T centerPoint;
+    private CenteredIntervalTree<T> leftTree; // contains all the intervals *completely* to the left of the center point
+    private CenteredIntervalTree<T> rightTree; // contains all the intervals *completely* to the right of the center point
+    private List<Interval<T>> intersectionsByStart = new LinkedList<>(); // intervals intersecting center point, sorted by start point
+    private List<Interval<T>> intersectionsByEnd = new LinkedList<>(); // ..., sorted by end point
+
+    public CenteredIntervalTree(T centerPoint, List<Interval<T>> intervals) {
+        this.centerPoint = centerPoint;
+        compute(intervals);
+    }
+
+    public CenteredIntervalTree(List<Interval<T>> intervals) {
+        this.centerPoint = intervals.get(0).getEnd(); // TODO replace this with something, its arbitrary
```

Comment:
This crashes for an empty list. Maybe we should allow an interval tree to contain nothing or make the constructor private and create a public static method that returns null in this case. Another option would be to check it here and throw an IllegalArgumentException instead of just crashing with an IndexOutOfBoundsException. We should also add a test that checks the behavior.

We should also throw an IllegalArgumentException if intervals is null (also with a test)

Reply:
I think a very simple improvement would be to take a random element to avoid worst case instances.

Reply:
A even worse case can be a null list passed in. If permitted by the project holder, we could make use of some widely used libraries for example Apache Commons Collections to complete null or empty check.